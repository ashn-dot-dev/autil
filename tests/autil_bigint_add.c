#define AUTIL_IMPLEMENTATION
#include "../autil.h"

struct autil_bigint* RES = NULL;
struct autil_bigint* LHS = NULL;
struct autil_bigint* RHS = NULL;
char* CSTR = NULL;

static void
test_autil_bigint_add__0_add_x(void)
{
    puts(__func__);
    RES = autil_bigint_new(AUTIL_BIGINT_ZERO);
    autil_bigint_add(RES, AUTIL_BIGINT_ZERO, AUTIL_BIGINT_POS_ONE);
    autil_bigint_dump(RES);
    autil_bigint_del(RES);

    RES = autil_bigint_new(AUTIL_BIGINT_ZERO);
    autil_bigint_add(RES, AUTIL_BIGINT_ZERO, AUTIL_BIGINT_NEG_ONE);
    autil_bigint_dump(RES);
    autil_bigint_del(RES);

    RES = autil_bigint_new(AUTIL_BIGINT_ZERO);
    RHS = autil_bigint_new_cstr("0x1234123412341234");
    autil_bigint_add(RES, AUTIL_BIGINT_ZERO, RHS);
    autil_bigint_dump(RES);
    autil_bigint_del(RES);
    autil_bigint_del(RHS);
}

static void
test_autil_bigint_add__x_add_0(void)
{
    puts(__func__);

    RES = autil_bigint_new(AUTIL_BIGINT_ZERO);
    autil_bigint_add(RES, AUTIL_BIGINT_POS_ONE, AUTIL_BIGINT_ZERO);
    autil_bigint_dump(RES);
    autil_bigint_del(RES);

    RES = autil_bigint_new(AUTIL_BIGINT_ZERO);
    autil_bigint_add(RES, AUTIL_BIGINT_NEG_ONE, AUTIL_BIGINT_ZERO);
    autil_bigint_dump(RES);
    autil_bigint_del(RES);

    RES = autil_bigint_new(AUTIL_BIGINT_ZERO);
    LHS = autil_bigint_new_cstr("0x1234123412341234");
    autil_bigint_add(RES, LHS, AUTIL_BIGINT_ZERO);
    autil_bigint_dump(RES);
    autil_bigint_del(RES);
    autil_bigint_del(LHS);
}

static void
test_autil_bigint_add__pos_add_pos(void)
{
    puts(__func__);

    RES = autil_bigint_new(AUTIL_BIGINT_ZERO);
    autil_bigint_add(RES, AUTIL_BIGINT_POS_ONE, AUTIL_BIGINT_POS_ONE);
    autil_bigint_dump(RES);
    autil_bigint_del(RES);

    LHS = autil_bigint_new_cstr("0x3210321032103210");
    RHS = autil_bigint_new_cstr("0x4567456745674567");
    RES = autil_bigint_new(AUTIL_BIGINT_ZERO);
    autil_bigint_add(RES, LHS, RHS);
    autil_bigint_dump(RES);
    autil_bigint_del(RES);
    autil_bigint_del(LHS);
    autil_bigint_del(RHS);
}

static void
test_autil_bigint_add__neg_add_neg(void)
{
    puts(__func__);

    RES = autil_bigint_new(AUTIL_BIGINT_ZERO);
    autil_bigint_add(RES, AUTIL_BIGINT_NEG_ONE, AUTIL_BIGINT_NEG_ONE);
    autil_bigint_dump(RES);
    autil_bigint_del(RES);

    LHS = autil_bigint_new_cstr("-0x3210321032103210");
    RHS = autil_bigint_new_cstr("-0x4567456745674567");
    RES = autil_bigint_new(AUTIL_BIGINT_ZERO);
    autil_bigint_add(RES, LHS, RHS);
    autil_bigint_dump(RES);
    autil_bigint_del(RES);
    autil_bigint_del(LHS);
    autil_bigint_del(RHS);
}

static void
test_autil_bigint_add__pos_add_neg(void)
{
    puts(__func__);

    RES = autil_bigint_new(AUTIL_BIGINT_ZERO);
    autil_bigint_add(RES, AUTIL_BIGINT_POS_ONE, AUTIL_BIGINT_NEG_ONE);
    autil_bigint_dump(RES);
    autil_bigint_del(RES);

    LHS = autil_bigint_new_cstr("+0x4567456745674567");
    RHS = autil_bigint_new_cstr("-0x3210321032103210");
    RES = autil_bigint_new(AUTIL_BIGINT_ZERO);
    autil_bigint_add(RES, LHS, RHS);
    autil_bigint_dump(RES);
    autil_bigint_del(RES);
    autil_bigint_del(LHS);
    autil_bigint_del(RHS);
}

static void
test_autil_bigint_add__neg_add_pos(void)
{
    puts(__func__);
    struct autil_bigint* INT = NULL;

    INT = autil_bigint_new(AUTIL_BIGINT_ZERO);
    autil_bigint_add(INT, AUTIL_BIGINT_NEG_ONE, AUTIL_BIGINT_POS_ONE);
    autil_bigint_dump(INT);
    autil_bigint_del(INT);

    LHS = autil_bigint_new_cstr("-0x4567456745674567");
    RHS = autil_bigint_new_cstr("+0x3210321032103210");
    RES = autil_bigint_new(AUTIL_BIGINT_ZERO);
    autil_bigint_add(RES, LHS, RHS);
    autil_bigint_dump(RES);
    autil_bigint_del(RES);
    autil_bigint_del(LHS);
    autil_bigint_del(RHS);
}

static void
test_autil_bigint_add__misc(void)
{
    puts(__func__);

    LHS = autil_bigint_new_cstr("+0xffffffffffffffffffffffffffffffff");
    RHS = autil_bigint_new_cstr("+0xffffffffffffffffffffffffffffffff");
    RES = autil_bigint_new(AUTIL_BIGINT_ZERO);
    autil_bigint_add(RES, LHS, RHS);
    CSTR = autil_bigint_to_new_cstr(RES, "#+x");
    puts(CSTR);
    autil_bigint_del(RES);
    autil_bigint_del(LHS);
    autil_bigint_del(RHS);
    autil_xalloc(CSTR, AUTIL_XALLOC_FREE);

    LHS = autil_bigint_new_cstr(
        "+0x"
        "c9fe78519191bc2978d99a79a3fcc5cc54fd2967b4b2684c1dcbbb6682b6aa53");
    RHS = autil_bigint_new_cstr(
        "-0x"
        "78b9091631d4444d2c12fb0d35ef6185e9bdf454c842afc5891a18616df60f09");
    RES = autil_bigint_new(AUTIL_BIGINT_ZERO);
    autil_bigint_add(RES, LHS, RHS);
    CSTR = autil_bigint_to_new_cstr(RES, "#+x");
    puts(CSTR);
    autil_bigint_del(RES);
    autil_bigint_del(LHS);
    autil_bigint_del(RHS);
    autil_xalloc(CSTR, AUTIL_XALLOC_FREE);
}

int
main(void)
{
    test_autil_bigint_add__0_add_x();
    test_autil_bigint_add__x_add_0();

    test_autil_bigint_add__pos_add_pos();
    test_autil_bigint_add__neg_add_neg();
    test_autil_bigint_add__pos_add_neg();
    test_autil_bigint_add__neg_add_pos();

    test_autil_bigint_add__misc();

    return EXIT_SUCCESS;
}
