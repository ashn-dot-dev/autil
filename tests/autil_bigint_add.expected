test_autil_bigint_add__0_add_x
SIGN: '+', COUNT: 1, LIMBS: [0x01]
SIGN: '-', COUNT: 1, LIMBS: [0x01]
SIGN: '+', COUNT: 8, LIMBS: [0x34, 0x12, 0x34, 0x12, 0x34, 0x12, 0x34, 0x12]
test_autil_bigint_add__x_add_0
SIGN: '+', COUNT: 1, LIMBS: [0x01]
SIGN: '-', COUNT: 1, LIMBS: [0x01]
SIGN: '+', COUNT: 8, LIMBS: [0x34, 0x12, 0x34, 0x12, 0x34, 0x12, 0x34, 0x12]
test_autil_bigint_add__pos_add_pos
SIGN: '+', COUNT: 1, LIMBS: [0x02]
SIGN: '+', COUNT: 8, LIMBS: [0x77, 0x77, 0x77, 0x77, 0x77, 0x77, 0x77, 0x77]
test_autil_bigint_add__neg_add_neg
SIGN: '-', COUNT: 1, LIMBS: [0x02]
SIGN: '-', COUNT: 8, LIMBS: [0x77, 0x77, 0x77, 0x77, 0x77, 0x77, 0x77, 0x77]
test_autil_bigint_add__pos_add_neg
SIGN: '0', COUNT: 0, LIMBS: []
SIGN: '+', COUNT: 8, LIMBS: [0x57, 0x13, 0x57, 0x13, 0x57, 0x13, 0x57, 0x13]
test_autil_bigint_add__neg_add_pos
SIGN: '0', COUNT: 0, LIMBS: []
SIGN: '-', COUNT: 8, LIMBS: [0x57, 0x13, 0x57, 0x13, 0x57, 0x13, 0x57, 0x13]
test_autil_bigint_add__misc
+0x1fffffffffffffffffffffffffffffffe
+0x51456f3b5fbd77dc4cc69f6c6e0d64466b3f3512ec6fb88694b1a30514c09b4a
